consensus/fsm/fsm.go

```diff
@@ -0,0 +1,499 @@
+}
+
+// ProduceReceiveLockEndorsementEvent produces an eReceiveLockEndorsement event right away
+func (m *ConsensusFSM) ProduceReceiveLockEndorsementEvent(endorsement Endorsement) {
+	m.produce(m.ctx.NewConsensusEvent(eReceiveLockEndorsement, endorsement), 0)
+}
+
+// ProduceReceivePreCommitEndorsementEvent produces an eReceivePreCommitEndorsement event right away
+func (m *ConsensusFSM) ProduceReceivePreCommitEndorsementEvent(endorsement Endorsement) {
+	m.produce(m.ctx.NewConsensusEvent(eReceivePreCommitEndorsement, endorsement), 0)
+}
+
+func (m *ConsensusFSM) produceConsensusEvent(et fsm.EventType, delay time.Duration) {
+	m.produce(m.ctx.NewConsensusEvent(et, nil), delay)
+}
+
+// produce adds an event into the queue for the consensus FSM to process
+func (m *ConsensusFSM) produce(evt *ConsensusEvent, delay time.Duration) {
+	consensusMtc.WithLabelValues(string(evt.Type())).Inc()
+	if delay > 0 {
+		m.wg.Add(1)
+		go func() {
+			select {
+			case <-m.close:
+			case <-m.clock.After(delay):
+				m.evtq <- evt
+			}
+			m.wg.Done()
+		}()
+	} else {
+		m.evtq <- evt
+	}
+}
+
+func (m *ConsensusFSM) prepare(_ fsm.Event) (fsm.State, error) {
+	delay, err := m.ctx.Prepare()
+	if err != nil || !m.ctx.IsDelegate() {
+		m.ProducePrepareEvent(delay)
+		return sPrepare, err
+	}
+	if delay > 0 {
+		time.Sleep(delay)
+	}
+	m.ctx.Logger().Info("Start a new round", zap.Int64("ts", m.clock.Now().Unix()))
+	// Setup timeout for waiting for proposed block
+	ttl := m.cfg.AcceptBlockTTL
+	m.produceConsensusEvent(eFailedToReceiveBlock, ttl)
+	ttl += m.cfg.AcceptProposalEndorsementTTL
+	m.produceConsensusEvent(eStopReceivingProposalEndorsement, ttl)
+	ttl += m.cfg.AcceptLockEndorsementTTL
+	m.produceConsensusEvent(eStopReceivingLockEndorsement, ttl)
+	m.produceConsensusEvent(eFailedToReachConsensusInTime, m.cfg.ProposerInterval)
+	// TODO add timeout for commit collection
+	if m.ctx.IsProposer() {
+		m.ctx.Logger().Info("current node is the proposer")
+		blk, err := m.ctx.MintBlock()
+		if err != nil {
+			// TODO: review the return state
+			return sPrepare, errors.Wrap(err, "error when minting a block")
+		}
+		m.ctx.Logger().Info("Broadcast init proposal.", log.Hex("blockHash", blk.Hash()))
+		m.ProduceReceiveBlockEvent(blk)
+		m.ctx.BroadcastBlockProposal(blk)
+	}
+
+	return sAcceptBlockProposal, nil
+}
+
+func (m *ConsensusFSM) onReceiveBlock(evt fsm.Event) (fsm.State, error) {
+	m.ctx.Logger().Debug("Receive block", zap.Int64("ts", m.clock.Now().Unix()))
+	cEvt, ok := evt.(*ConsensusEvent)
+	if !ok {
+		return sAcceptBlockProposal, errors.Wrapf(ErrEvtCast, "invalid fsm event %+v", evt)
+	}
+	block, ok := cEvt.Data().(Endorsement)
+	if !ok {
+		return sAcceptBlockProposal, errors.Wrapf(ErrEvtConvert, "invalid data type")
+	}
+	en, err := m.ctx.NewProposalEndorsement(block)
+	if err != nil {
+		return sAcceptBlockProposal, errors.Wrap(err, "failed to generate proposal endorsement")
```

Comment:
No need to return error (at least not always), right?

Block validation failure should be classified into an error of state transition